libs/pika/executors/tests/unit/limiting_scheduler.cpp

```diff
@@ -0,0 +1,259 @@
+#include <pika/executors/inline_scheduler.hpp>
+#include <pika/executors/limiting_scheduler.hpp>
+#include <pika/init.hpp>
+#include <pika/testing.hpp>
+
+#include <atomic>
+#include <chrono>
+#include <cstddef>
+#include <cstdint>
+#include <cstdlib>
+#include <random>
+#include <string>
+#include <utility>
+#include <vector>
+
+// this test launches many tasks continuously using a limiting scheduler
+// some of the tasks will suspend themselves randomly and so new tasks will be
+// spawned until the max tasks 'in flight' limit is reached. Once this happens
+// no new tasks should be created and the number 'in flight' should remain
+// below/equal to the limit. We track this using some counters and check that
+// at the end of the test, the max counter never exceeded the limit.
+
+// type for counters we will use to track tasks in flight
+using atomic_type = std::atomic<std::int64_t>;
+
+namespace ex = pika::execution::experimental;
+using namespace pika::debug::detail;
+
+///////////////////////////////////////////////////////////////////////////////
+/// \brief update_maximum atomically using CAS
+/// \param atomic maximum_value
+/// \param new value
+///
+template <typename T>
+void update_maximum(std::atomic<T>& maximum, T const& value) noexcept
+{
+    // if another thread changes the max, then this will either retry, and
+    // then succeed in updating, or fail if the other thread set a higher value
+    // (because on fail, prev is updated by CAS)
+    T prev = maximum;
+    while ((value > prev) && !maximum.compare_exchange_weak(prev, value))
+    {
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// random work
+void work_fn()
+{
+    std::default_random_engine eng;
+    std::uniform_int_distribution<std::size_t> idist(10, 50);
+    std::size_t loop = idist(eng);
+    for (std::size_t i = 0; i < loop * 1000; ++i)
+    {
+        volatile double x;
+        x = std::log10(i);
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+//  simple task that can yield at random and increments counters
+void test_fn(atomic_type& active_count, atomic_type& total_count, atomic_type& active_max)
+{
+    // Increment active task count and max active tasks atomically
+    update_maximum(active_max, ++active_count);
+    ++total_count;
+
+    // yield some random amount of times to make the test more realistic
+    // this allows other tasks to run and tests if the limiting scheduler
+    // is doing its job
+    std::default_random_engine eng;
+    std::uniform_int_distribution<std::size_t> idist(10, 20);
+    std::size_t loop = idist(eng);
+    for (std::size_t i = 0; i < loop; ++i)
+    {
+        pika::this_thread::yield();
+        work_fn();
+    }
+
+    // task is completing, decrement active task count
+    --active_count;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+void test_limit_inline()
```

Comment:
Could you not template this on the scheduler? Or are there important differences in the two tests that you can't combine into one function?

Reply:
Only real difference is that the inline one tests that the tasks have the same ID, but the other one has different IDs (TEST_NEQ/TEST_EQ)